// @ant-design/pro-provider/lib/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.trTRIntl = exports.srRSIntl = exports.ruRUIntl = exports.ptBRIntl = exports.plPLIntl = exports.msMYIntl = exports.koKRIntl = exports.jaJPIntl = exports.itITIntl = exports.intlMapKeys = exports.intlMap = exports.idIDIntl = exports.frFRIntl = exports.faIRIntl = exports.esESIntl = exports.enUSIntl = exports.enGBIntl = exports.default = exports.deDEIntl = exports.createIntl = exports.caESIntl = exports.arEGIntl = exports.ConfigProviderWrap = exports.ConfigProvider = exports.ConfigConsumer = void 0;
exports.useIntl = useIntl;
exports.zhTWIntl = exports.zhCNIntl = exports.viVNIntl = void 0;

require("antd/es/config-provider/style");

var _configProvider = _interopRequireDefault(require("antd/es/config-provider"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _createForOfIteratorHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/createForOfIteratorHelper"));

var _react = _interopRequireWildcard(require("react"));

var _swr = require("swr");

var _zh_CN = _interopRequireDefault(require("antd/lib/locale/zh_CN"));

var _ar_EG = _interopRequireDefault(require("./locale/ar_EG"));

var _zh_CN2 = _interopRequireDefault(require("./locale/zh_CN"));

var _en_US = _interopRequireDefault(require("./locale/en_US"));

var _en_GB = _interopRequireDefault(require("./locale/en_GB"));

var _vi_VN = _interopRequireDefault(require("./locale/vi_VN"));

var _it_IT = _interopRequireDefault(require("./locale/it_IT"));

var _es_ES = _interopRequireDefault(require("./locale/es_ES"));

var _ca_ES = _interopRequireDefault(require("./locale/ca_ES"));

var _ja_JP = _interopRequireDefault(require("./locale/ja_JP"));

var _ru_RU = _interopRequireDefault(require("./locale/ru_RU"));

var _sr_RS = _interopRequireDefault(require("./locale/sr_RS"));

var _ms_MY = _interopRequireDefault(require("./locale/ms_MY"));

var _zh_TW = _interopRequireDefault(require("./locale/zh_TW"));

var _fr_FR = _interopRequireDefault(require("./locale/fr_FR"));

var _pt_BR = _interopRequireDefault(require("./locale/pt_BR"));

var _ko_KR = _interopRequireDefault(require("./locale/ko_KR"));

var _id_ID = _interopRequireDefault(require("./locale/id_ID"));

var _de_DE = _interopRequireDefault(require("./locale/de_DE"));

var _fa_IR = _interopRequireDefault(require("./locale/fa_IR"));

var _tr_TR = _interopRequireDefault(require("./locale/tr_TR"));

var _pl_PL = _interopRequireDefault(require("./locale/pl_PL"));

function get(source, path, defaultValue) {
  // a[3].b -> a.3.b
  var paths = path.replace(/\[(\d+)\]/g, '.$1').split('.');
  var result = source;
  var message = defaultValue; // eslint-disable-next-line no-restricted-syntax

  var _iterator = (0, _createForOfIteratorHelper2.default)(paths),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var p = _step.value;
      message = Object(result)[p];
      result = Object(result)[p];

      if (message === undefined) {
        return defaultValue;
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return message;
}
/**
 * 创建一个操作函数
 *
 * @param locale
 * @param localeMap
 */


var createIntl = function createIntl(locale, localeMap) {
  return {
    getMessage: function getMessage(id, defaultMessage) {
      return get(localeMap, id, defaultMessage) || defaultMessage;
    },
    locale: locale
  };
};

exports.createIntl = createIntl;
var arEGIntl = createIntl('ar_EG', _ar_EG.default);
exports.arEGIntl = arEGIntl;
var zhCNIntl = createIntl('zh_CN', _zh_CN2.default);
exports.zhCNIntl = zhCNIntl;
var enUSIntl = createIntl('en_US', _en_US.default);
exports.enUSIntl = enUSIntl;
var enGBIntl = createIntl('en_GB', _en_GB.default);
exports.enGBIntl = enGBIntl;
var viVNIntl = createIntl('vi_VN', _vi_VN.default);
exports.viVNIntl = viVNIntl;
var itITIntl = createIntl('it_IT', _it_IT.default);
exports.itITIntl = itITIntl;
var jaJPIntl = createIntl('ja_JP', _ja_JP.default);
exports.jaJPIntl = jaJPIntl;
var esESIntl = createIntl('es_ES', _es_ES.default);
exports.esESIntl = esESIntl;
var caESIntl = createIntl('ca_ES', _ca_ES.default);
exports.caESIntl = caESIntl;
var ruRUIntl = createIntl('ru_RU', _ru_RU.default);
exports.ruRUIntl = ruRUIntl;
var srRSIntl = createIntl('sr_RS', _sr_RS.default);
exports.srRSIntl = srRSIntl;
var msMYIntl = createIntl('ms_MY', _ms_MY.default);
exports.msMYIntl = msMYIntl;
var zhTWIntl = createIntl('zh_TW', _zh_TW.default);
exports.zhTWIntl = zhTWIntl;
var frFRIntl = createIntl('fr_FR', _fr_FR.default);
exports.frFRIntl = frFRIntl;
var ptBRIntl = createIntl('pt_BR', _pt_BR.default);
exports.ptBRIntl = ptBRIntl;
var koKRIntl = createIntl('ko_KR', _ko_KR.default);
exports.koKRIntl = koKRIntl;
var idIDIntl = createIntl('id_ID', _id_ID.default);
exports.idIDIntl = idIDIntl;
var deDEIntl = createIntl('de_DE', _de_DE.default);
exports.deDEIntl = deDEIntl;
var faIRIntl = createIntl('fa_IR', _fa_IR.default);
exports.faIRIntl = faIRIntl;
var trTRIntl = createIntl('tr_TR', _tr_TR.default);
exports.trTRIntl = trTRIntl;
var plPLIntl = createIntl('pl_PL', _pl_PL.default);
exports.plPLIntl = plPLIntl;
var intlMap = {
  'ar-EG': arEGIntl,
  'zh-CN': zhCNIntl,
  'en-US': enUSIntl,
  'en-GB': enGBIntl,
  'vi-VN': viVNIntl,
  'it-IT': itITIntl,
  'ja-JP': jaJPIntl,
  'es-ES': esESIntl,
  'ca-ES': caESIntl,
  'ru-RU': ruRUIntl,
  'sr-RS': srRSIntl,
  'ms-MY': msMYIntl,
  'zh-TW': zhTWIntl,
  'fr-FR': frFRIntl,
  'pt-BR': ptBRIntl,
  'ko-KR': koKRIntl,
  'id-ID': idIDIntl,
  'de-DE': deDEIntl,
  'fa-IR': faIRIntl,
  'tr-TR': trTRIntl,
  'pl-PL': plPLIntl
};
exports.intlMap = intlMap;
var intlMapKeys = Object.keys(intlMap);
exports.intlMapKeys = intlMapKeys;

var ConfigContext = /*#__PURE__*/_react.default.createContext({
  intl: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, zhCNIntl), {}, {
    locale: 'default'
  }),
  valueTypeMap: {}
});

var ConfigConsumer = ConfigContext.Consumer,
    ConfigProvider = ConfigContext.Provider;
/**
 * 根据 antd 的 key 来找到的 locale 插件的 key
 *
 * @param localeKey
 */

exports.ConfigProvider = ConfigProvider;
exports.ConfigConsumer = ConfigConsumer;

var findIntlKeyByAntdLocaleKey = function findIntlKeyByAntdLocaleKey(localeKey) {
  if (!localeKey) {
    return 'zh-CN';
  }

  var localeName = localeKey.toLocaleLowerCase();
  return intlMapKeys.find(function (intlKey) {
    var LowerCaseKey = intlKey.toLocaleLowerCase();
    return LowerCaseKey.includes(localeName);
  });
};
/**
 * 组件解除挂载后清空一下 cache
 *
 * @returns
 */


var CacheClean = function CacheClean() {
  var _useSWRConfig = (0, _swr.useSWRConfig)(),
      cache = _useSWRConfig.cache;

  (0, _react.useEffect)(function () {
    return function () {
      // is a map
      // @ts-ignore
      cache.clear();
    }; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return null;
};
/**
 * 如果没有配置 locale，这里组件会根据 antd 的 key 来自动选择
 *
 * @param param0
 */


var ConfigProviderWrap = function ConfigProviderWrap(_ref) {
  var children = _ref.children,
      _ref$autoClearCache = _ref.autoClearCache,
      autoClearCache = _ref$autoClearCache === void 0 ? false : _ref$autoClearCache;

  var _useContext = (0, _react.useContext)(_configProvider.default.ConfigContext),
      locale = _useContext.locale; // 如果 locale 不存在自动注入的 AntdConfigProvider


  var Provider = locale === undefined ? _configProvider.default : _react.default.Fragment;

  var configProviderDom = /*#__PURE__*/_react.default.createElement(ConfigConsumer, null, function (value) {
    var _value$intl;

    var localeName = locale === null || locale === void 0 ? void 0 : locale.locale;
    var key = findIntlKeyByAntdLocaleKey(localeName); // antd 的 key 存在的时候以 antd 的为主

    var intl = localeName && ((_value$intl = value.intl) === null || _value$intl === void 0 ? void 0 : _value$intl.locale) === 'default' ? intlMap[key] : value.intl || intlMap[key]; // 自动注入 antd 的配置

    var configProvider = locale === undefined ? {
      locale: _zh_CN.default
    } : {};
    return /*#__PURE__*/_react.default.createElement(Provider, configProvider, /*#__PURE__*/_react.default.createElement(ConfigProvider, {
      value: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, value), {}, {
        intl: intl || zhCNIntl
      })
    }, autoClearCache && /*#__PURE__*/_react.default.createElement(CacheClean, null), children));
  });

  if (!autoClearCache) return configProviderDom;
  return /*#__PURE__*/_react.default.createElement(_swr.SWRConfig, {
    value: {
      provider: function provider() {
        return new Map();
      }
    }
  }, configProviderDom);
};

exports.ConfigProviderWrap = ConfigProviderWrap;

function useIntl() {
  var context = (0, _react.useContext)(ConfigContext);
  return context.intl || zhCNIntl;
}

var _default = ConfigContext;
exports.default = _default;
