// react-css-variables/lib/index.js
'use strict';

exports.__esModule = true;

exports.default = function () {
  for (var _len = arguments.length, varNames = Array(_len), _key = 0; _key < _len; _key++) {
    varNames[_key] = arguments[_key];
  }

  return function (WrappedComponent) {
    return function (_Component) {
      _inherits(_class2, _Component);

      function _class2(props) {
        _classCallCheck(this, _class2);

        var _this = _possibleConstructorReturn(this, _Component.call(this, props));

        _this.displayName = (0, _utils.getDisplayName)(WrappedComponent);


        _this.state = {
          varProps: _lodash2.default.apply(undefined, [props].concat(varNames)),
          ownProps: _lodash4.default.apply(undefined, [props].concat(varNames))
        };
        return _this;
      }

      _class2.prototype.componentDidMount = function componentDidMount() {
        (0, _utils.setVariables)((0, _reactDom.findDOMNode)(this), this.state.varProps);
      };

      _class2.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
        var _this2 = this;

        var prevVarProps = this.state.varProps;

        this.setState({
          varProps: _lodash2.default.apply(undefined, [nextProps].concat(varNames)),
          ownProps: _lodash4.default.apply(undefined, [nextProps].concat(varNames))
        }, function () {

          if (!(0, _utils.shallowEqual)(prevVarProps, _this2.state.varProps)) {
            (0, _utils.setVariables)((0, _reactDom.findDOMNode)(_this2), _this2.state.varProps);
          }
        });
      };

      _class2.prototype.shouldComponentUpdate = function shouldComponentUpdate(nextProps, nextState) {
        return !(0, _utils.shallowEqual)(this.state.ownProps, nextState.ownProps);
      };

      _class2.prototype.render = function render() {
        return _react2.default.createElement(WrappedComponent, this.state.ownProps);
      };

      return _class2;
    }(_react.Component);
  };
};

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _lodash = require('lodash.pick');

var _lodash2 = _interopRequireDefault(_lodash);

var _lodash3 = require('lodash.omit');

var _lodash4 = _interopRequireDefault(_lodash3);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

module.exports = exports['default'];
